File: prototype/rails/src/app/domain/orders/order_payment.rb
require_relative "order_status"

module Domain
  module Orders
    module OrderPayment
      module_function

      # An order awaits a card for as long as one could still carry it to paid,
      # which is what the storefront asks before it shows a card form.
      def awaits_card?(status)
        OrderStatus.can_transition?(status, OrderStatus::PAID)
      end

      # A guest's order is unpaid before it is chargeable: verifying the address
      # is the step between.
      def unpaid?(status)
        awaits_card?(status) || OrderStatus.can_transition?(status, OrderStatus::AWAITING_PAYMENT)
      end

      def payable?(status, purchaser_verified)
        purchaser_verified && awaits_card?(status)
      end
    end
  end
end

File: prototype/node/src/app/core/orders/order-payment.ts
import { canTransitionOrder, type OrderStatus } from './order-status.ts'

/** An order awaits a card for as long as one could still carry it to paid,
 * which is what the storefront asks before it shows a card form. */
export function awaitsCard(status: OrderStatus): boolean {
  return canTransitionOrder(status, 'paid')
}

/** A guest's order is unpaid before it is chargeable: verifying the address
 * is the step between. */
export function isUnpaid(status: OrderStatus): boolean {
  return awaitsCard(status) || canTransitionOrder(status, 'awaiting_payment')
}

export function isPayable(status: OrderStatus, isPurchaserVerified: boolean): boolean {
  return isPurchaserVerified && awaitsCard(status)
}
